async/kafka/group/group.go

```diff
@@ -0,0 +1,179 @@
+type Factory struct {
+	name    string
+	group   string
+	topic   string
+	brokers []string
+	oo      []kafka.OptionFunc
+}
+
+// New constructor.
+func New(name, group, topic string, brokers []string, oo ...kafka.OptionFunc) (*Factory, error) {
+
+	if name == "" {
+		return nil, errors.New("name is required")
+	}
+
+	if group == "" {
+		return nil, errors.New("group is required")
+	}
+
+	if len(brokers) == 0 {
+		return nil, errors.New("provide at least one broker")
+	}
+
+	if topic == "" {
+		return nil, errors.New("topic is required")
+	}
+
+	return &Factory{name: name, group: group, topic: topic, brokers: brokers, oo: oo}, nil
+}
+
+// Create a new consumer.
+func (f *Factory) Create() (async.Consumer, error) {
+
+	config, err := kafka.SaramaConfig(f.name)
+
+	if err != nil {
+		return nil, err
+	}
+
+	cc := kafka.ConsumerConfig{
+		Brokers:      f.brokers,
+		Buffer:       0,
+		SaramaConfig: config,
+	}
+
+	c := &consumer{
+		topic:       f.topic,
+		group:       f.group,
+		traceTag:    opentracing.Tag{Key: "group", Value: f.group},
+		consumerCnf: cc,
+	}
+
+	for _, o := range f.oo {
+		err = o(c)
+		if err != nil {
+			return nil, fmt.Errorf("Could not apply OptionFunc '%v' to consumer : %v", runtime.FuncForPC(reflect.ValueOf(o).Pointer()).Name(), err)
+		}
+	}
+
+	return c, nil
+}
+
+// consumer members can be injected or overwritten with the usage of OptionFunc arguments.
+type consumer struct {
+	kafka.Consumer
+	topic       string
+	group       string
+	traceTag    opentracing.Tag
+	cnl         context.CancelFunc
+	cg          sarama.ConsumerGroup
+	consumerCnf kafka.ConsumerConfig
+}
+
+func (c *consumer) ConsumerConfig() *kafka.ConsumerConfig { return &c.consumerCnf }
+
+// Close handles closing consumer.
+func (c *consumer) Close() error {
+	if c.cnl != nil {
+		c.cnl()
+	}
+
+	return errors.Wrap(c.cg.Close(), "failed to close consumer")
+}
+
+// Consume starts consuming messages from a Kafka topic.
+func (c *consumer) Consume(ctx context.Context) (<-chan async.Message, <-chan error, error) {
+	ctx, cnl := context.WithCancel(ctx)
+	c.cnl = cnl
+
+	return consumeWithGroup(ctx, c)
+}
+
+func consumeWithGroup(ctx context.Context, c *consumer) (<-chan async.Message, <-chan error, error) {
+
+	cg, err := sarama.NewConsumerGroup(c.ConsumerConfig().Brokers, c.group, c.ConsumerConfig().SaramaConfig)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create consumer")
+	}
+	c.cg = cg
+	log.Infof("consuming messages from topic '%s' using group '%s'", c.topic, c.group)
+
+	chMsg := make(chan async.Message, c.ConsumerConfig().Buffer)
+	chErr := make(chan error, c.ConsumerConfig().Buffer)
+
+	go func(consumer sarama.ConsumerGroup) {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info("canceling consuming messages requested")
+				closeConsumer(consumer)
+				return
+			case consumerError := <-consumer.Errors():
+				closeConsumer(consumer)
+				chErr <- consumerError
+				return
+			}
+		}
+	}(c.cg)
```

Comment:
This is safe but since `c.cg` does not change we can omit it. If it was the value of a loop then we should capture itthat way.